DigitalLearningSolutions.Data/Services/UserService.cs

```diff
@@ -193,77 +201,75 @@ List<DelegateUser> delegateUsers
+                centreAnswersData.CentreEmail
+            );
 
-            if (verifiedDelegateUsers.Count != 0)
+            if (delegateUsers.Count != 0)
             {
-                userDataService.UpdateUser(
-                    myAccountDetailsData.FirstName,
-                    myAccountDetailsData.Surname,
-                    myAccountDetailsData.Email,
-                    myAccountDetailsData.ProfileImage,
-                    myAccountDetailsData.ProfessionalRegistrationNumber,
-                    myAccountDetailsData.HasBeenPromptedForPrn,
-                    centreAnswersData.JobGroupId,
-                    1 // TODO HEEDLS-887 This needs correcting to the correct UserId for the delegate record.
+                var delegateId = myAccountDetailsData.DelegateId!.Value;
+                // TODO HEEDLS-887 Remove this call once groupsService.SynchroniseUserChangesWithGroups is refactored
+                var oldDelegateUserDetails = userDataService.GetDelegateUserById(delegateId);
+
+                userDataService.UpdateDelegateUserCentrePrompts(
+                    delegateId,
+                    centreAnswersData.Answer1,
+                    centreAnswersData.Answer2,
+                    centreAnswersData.Answer3,
+                    centreAnswersData.Answer4,
+                    centreAnswersData.Answer5,
+                    centreAnswersData.Answer6,
+                    detailsLastChecked
                 );
 
-                var oldDelegateDetails =
-                    verifiedDelegateUsers.SingleOrDefault(u => u.Id == myAccountDetailsData.DelegateId);
-
-                if (oldDelegateDetails != null && centreAnswersData != null)
-                {
-                    userDataService.UpdateDelegateUserCentrePrompts(
-                        myAccountDetailsData.DelegateId!.Value,
-                        centreAnswersData.Answer1,
-                        centreAnswersData.Answer2,
-                        centreAnswersData.Answer3,
-                        centreAnswersData.Answer4,
-                        centreAnswersData.Answer5,
-                        centreAnswersData.Answer6
-                    );
-
-                    groupsService.SynchroniseUserChangesWithGroups(
-                        oldDelegateDetails,
-                        myAccountDetailsData,
-                        centreAnswersData
-                    );
-                }
+                groupsService.SynchroniseUserChangesWithGroups(
+                    oldDelegateUserDetails!,
+                    myAccountDetailsData,
+                    centreAnswersData
+                );
             }
         }
 
-        public bool NewEmailAddressIsValid(string emailAddress, int? adminUserId, int? delegateUserId, int centreId)
+        public bool NewEmailAddressIsValid(
+            string emailAddress,
+            int? adminUserId,
+            int? delegateUserId,
+            int? userId = null

```

Comment:
I'm a bit scared that this method is allowing people to pass in adminId and delegateId and userId.

What happens if someone decides to call this method with a userId and a delegateId which correspond to different users?

We could avoid this by disallowing any calling code from passing in adminUserId or delegateUserId, demanding that it gives us a userId and centreId instead (which makes the implementation of this method simpler).
Can we do this instead?

Reply:
I don't think I need the centre id, I can just return `!userDataService.EmailIsInUseByOtherUser(userId, emailAddress)`, or am I missing something?

Reply:
For checking UserCentreDetails.

Reply:
Use `userDataService.EmailIsInUseByOtherUser(userId, emailAddress)` instead of this method.